src/metabase/models/notification.clj

```diff
@@ -62,16 +84,55 @@
    #(group-by :notification_id
               (t2/select :model/NotificationHandler :notification_id [:in (map :id notifications)]))
    :id
-   {:default nil}))
+   {:default []}))
+
+(def ^:private Notification
+  [:merge
+   [:map
+    [:payload_type (into [:enum] notification-types)]]
+   [:multi {:dispatch (comp keyword :payload_type)}
+    [:notification/system-event
+     [:map
+      [:payload_id {:optional true} nil?]]]
+    [:notification/card
+     [:map
+      [:payload_id int?]
+      [:creator_id int?]]]
+    [:notification/testing :any]]])
+
+(defn- validate-notification
+  [notification]
+  (mu/validate-throw Notification notification))
+
+(t2/define-before-insert :model/Notification
+  [instance]
+  (validate-notification instance)
+  instance)
+
+(t2/define-before-update :model/Notification
+  [instance]
+  (validate-notification instance)
+  (when (some #{:payload_type :payload_id} (keys (t2/changes instance)))
+    (throw (ex-info "Update notification payload is not allowed."
+                    {:status-code 400
+                     :changes     (t2/changes instance)})))
+  instance)
 
 (defn- delete-trigger-for-subscription!
   [& args]
   (apply (requiring-resolve 'metabase.task.notification/delete-trigger-for-subscription!) args))
 
 (t2/define-before-delete :model/Notification
   [instance]
-  (doseq [subscription-ids (t2/select-pks-set :model/NotificationSubscription :notification_id (:id instance))]
-    (delete-trigger-for-subscription! subscription-ids))
+  (doseq [subscription-id (t2/select-pks-set :model/NotificationSubscription
+                                             :notification_id (:id instance)
+                                             :type :notification-subscription/cron)]
+    (delete-trigger-for-subscription! subscription-id))
+  (when-let [payload-id (:payload_id instance)]
+    (t2/delete! (case (:payload_type instance)
+                  :notification/card
+                  :model/NotificationCard)
```

Comment:
According to the `Notification` type above, there is an optional `:payload_id` for a `:notification/system-event`, but that would cause this case statement to throw.

Should it not be a no-op?

Reply:
for `system-event` payload_id should be `nil` so this branch should never reached

Reply:
Should we not update this schema then?

```clojure
 [:notification/system-event
     [:map
      [:payload_id {:optional true} nil?]]
```

Reply:
D'oh, didn't look closely enough to see that it had `nil?` as the type.